nf_core/components/update.py

```diff
@@ -0,0 +1,851 @@
+            patch_relpath (Path | str): The path to patch file in the pipeline
+            component_dir (Path | str): The module/subworkflow directory in the pipeline
+            component_install_dir (Path | str): The directory where the new component
+                                            file have been installed
+
+        Returns:
+            (bool): Whether the patch application was successful
+        """
+        component_fullname = str(Path(repo_path, component))
+        log.info(f"Found patch for  {self.component_type[:-1]} '{component_fullname}'. Trying to apply it to new files")
+
+        patch_path = Path(self.dir / patch_relpath)
+        component_relpath = Path(self.component_type, repo_path, component)
+
+        # Check that paths in patch file are updated
+        self.check_patch_paths(patch_path, component)
+
+        # Copy the installed files to a new temporary directory to save them for later use
+        temp_dir = Path(tempfile.mkdtemp())
+        temp_component_dir = temp_dir / component
+        shutil.copytree(component_install_dir, temp_component_dir)
+
+        try:
+            new_files = ModulesDiffer.try_apply_patch(component, repo_path, patch_path, temp_component_dir)
+        except LookupError:
+            # Patch failed. Save the patch file by moving to the install dir
+            shutil.move(patch_path, Path(component_install_dir, patch_path.relative_to(component_dir)))
+            log.warning(
+                f"Failed to apply patch for {self.component_type[:-1]} '{component_fullname}'. You will have to apply the patch manually"
+            )
+            return False
+
+        # Write the patched files to a temporary directory
+        log.debug("Writing patched files")
+        for file, new_content in new_files.items():
+            fn = temp_component_dir / file
+            with open(fn, "w") as fh:
+                fh.writelines(new_content)
+
+        # Create the new patch file
+        log.debug("Regenerating patch file")
+        ModulesDiffer.write_diff_file(
+            Path(temp_component_dir, patch_path.relative_to(component_dir)),
+            component,
+            repo_path,
+            component_install_dir,
+            temp_component_dir,
+            file_action="w",
+            for_git=False,
+            dsp_from_dir=component_relpath,
+            dsp_to_dir=component_relpath,
+        )
+
+        # Move the patched files to the install dir
+        log.debug("Overwriting installed files installed files  with patched files")
+        shutil.rmtree(component_install_dir)
+        shutil.copytree(temp_component_dir, component_install_dir)
+
+        # Add the patch file to the modules.json file
+        self.modules_json.add_patch_entry(
+            component, self.modules_repo.remote_url, repo_path, patch_relpath, write_file=write_file
+        )
+
+        return True
+
+    def get_modules_subworkflows_to_update(self, component, modules_repo):
```

Comment:
I have a similar method in https://github.com/nf-core/tools/pull/2012/files#diff-894c97c2057bb64eee0b66a2fc104f6ad29ed0efc38e7a2ea0d0f50504923eb1R918
should we move it to `utils.py`? For me `modules_json.py` makes sense for now.

Reply:
Sounds good :) We can put it in `modules_json.py` as it's parsing that file.

Reply:
thinking about it, yours goes actually the other direction of the dependency graph, so I think we should keep both.